packages/msgs-derive/src/lib.rs

```diff
@@ -568,50 +568,64 @@ impl Parse for AllPermissions {
-        #[allow(clippy::too_many_arguments)]
-        fn validate_external_contract<#(#fs),*, #(#cs),*>(
-                storage: &dyn cosmwasm_std::Storage,
-                contract_addr: Addr,
-                #(#contracts: #fs),*,
-                #(#contract_names: String),*
-            ) -> error_stack::Result<String, axelar_wasm_std::permission_control::Error>
-            where
-            #(#fs:FnOnce(&dyn cosmwasm_std::Storage) -> error_stack::Result<cosmwasm_std::Addr, #cs>),*,
-            #(#cs: error_stack::Context),*
-                {
-                #(
-                    match #contracts(storage) {
-                        Ok(stored_addr) => {
-                            if stored_addr == contract_addr {
-                                return Ok(#contract_names);
-                            }
-                        },
-                        Err(_) => {},
-                    }
-                )*
+        TokenStream::from(quote! {
+            // this function can be called with a lot of arguments, so we suppress the warning
+            #[allow(clippy::too_many_arguments)]
+            fn validate_external_contract<#(#fs),*, #(#cs),*>(
+                    storage: &dyn cosmwasm_std::Storage,
+                    contract_addr: Addr,
+                    #(#contracts: #fs),*,
+                    #(#contract_names: String),*
+                ) -> error_stack::Result<String, axelar_wasm_std::permission_control::Error>
+                where
+                #(#fs:FnOnce(&dyn cosmwasm_std::Storage) -> error_stack::Result<cosmwasm_std::Addr, #cs>),*,
+                #(#cs: error_stack::Context),*
+                    {
+                    #(
+                        match #contracts(storage) {
+                            Ok(stored_addr) => {
+                                if stored_addr == contract_addr {
+                                    return Ok(#contract_names);
+                                }
+                            },
+                            Err(_) => {},
+                        }
+                    )*
 
-            Err(error_stack::report!(axelar_wasm_std::permission_control::Error::Unauthorized))
-        }
-    })
+                Err(error_stack::report!(axelar_wasm_std::permission_control::Error::Unauthorized))
+            }
+        })
+    } else {
+        TokenStream::from(quote! {
+            fn validate_external_contract(
+                    storage: &dyn cosmwasm_std::Storage,
+                    contract_addr: Addr,
+                ) -> error_stack::Result<String, axelar_wasm_std::permission_control::Error>
+                    {
+                // This is only called when a relay message is executed. Since no proxy contract has
+                // permission to execute a message, this will always be an error.
+                Err(error_stack::report!(axelar_wasm_std::permission_control::Error::Unauthorized))
+            }
+        })
+    }
```

Comment:
try to nest as little as possible, and when you have large branch cases like this it's preferable to extract functions from them so it's more readable and you don't have to keep as much context in mind